File: src/repositories/UrlRepository.ts
import prisma from '../config/db';
import { Url } from '@prisma/client';

export class UrlRepository {
  async create(originalUrl: string, shortCode?: string): Promise<Url> {
    return prisma.url.create({
      data: {
        originalUrl,
        shortCode,
      },
    });
  }

  async updateShortCode(id: number, shortCode: string): Promise<Url> {
    return prisma.url.update({
        where: { id },
        data: { shortCode }
    });
  }

  async findByShortCode(shortCode: string): Promise<Url | null> {
    return prisma.url.findUnique({
      where: {
        shortCode,
      },
    });
  }

  async incrementClicks(shortCode: string): Promise<Url> {
    return prisma.url.update({
      where: {
        shortCode,
      },
      data: {
        clicks: {
          increment: 1,
        },
      },
    });
  }

  async findByOriginalUrl(originalUrl: string): Promise<Url | null> {
    return prisma.url.findFirst({
        where: {
            originalUrl
        }
    });
  }
}

File: src/main/java/com/portfolio/urlshortener/repository/UrlRepository.java
package com.portfolio.urlshortener.repository;

import com.portfolio.urlshortener.model.Url;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface UrlRepository extends JpaRepository<Url, Integer> {
    Optional<Url> findByOriginalUrl(String originalUrl);
    Optional<Url> findByShortCode(String shortCode);
    
    @Modifying
    @Query("UPDATE Url u SET u.clicks = u.clicks + 1 WHERE u.shortCode = :shortCode")
    void incrementClicks(String shortCode);
}
